### PR TITLE
multi: remove static default channel types instead select them based on lnd's verison

### PR DESCRIPTION
**Pull request by Roasbeef**
In this PR, we remove the defaults and the CLI and RPC level. We replace them with a new set of functional options that'll let us decide dynamically which features to use based on the lnd version. This lets us support older versions, while being able to activated newer features as we detect them based on what lnd offers. 

#### Pull Request Checklist
- [x] LndServices minimum version has been updated if new lnd apis/fields are
  used.

### CODE REVIEW

File: rpcserver.go [hunk 1]
```diff
@@ -1185,18 +1186,66 @@ func prepareAndSubmitOrder(ctx context.Context, o order.Order,
 	)
 }
 
+// lndVersionIsCompatible makes sure the detected lnd version is compatible
+// with our current version requirements. True is returned if the version is
+// compatible, and false otherwise.
+func lndVersionIsCompatible(actual *verrpc.Version,
```
Comment (positiveblue) [inline rpcserver.go:1192]: I have some dobuts here:

- Do we need to check something else beyond the lnd version (`featurebits`?)

- Do we need to check the `BuildTags`?

- May be worth moving this function to `"github.com/lightningnetwork/lnd/lnrpc/verrpc"` itself so we do not have to copy paste it in multiple projects.

- Here we assume that all versions will be backwards compatible, we need to keep this in mind in the future then.

File: order/rpc_parse.go [hunk 2]
```diff
@@ -19,10 +19,41 @@ import (
...
+}
+
 // ParseRPCOrder parses the incoming raw RPC order into the go native data
 // types used in the order struct.
 func ParseRPCOrder(version, leaseDuration uint32,
-	details *poolrpc.Order) (*Kit, error) {
+	details *poolrpc.Order, parseOpts ...OrderParseOption) (*Kit, error) {
+
+	opts := defaultParseOptions()
+	for _, newOpt := range parseOpts {
```
Comment (positiveblue) [inline order/rpc_parse.go:54]: nice

**Review by positiveblue — APPROVED** [review]
LGTM but I have some doubts that may be blocking

File: rpcserver.go [hunk 3]
```diff
@@ -1185,18 +1186,66 @@ func prepareAndSubmitOrder(ctx context.Context, o order.Order,
...
+	expected *verrpc.Version) bool {
+
+	// We need to check the versions parts sequentially as they are
+	// hierarchical.
+	if actual.AppMajor != expected.AppMajor {
```
Comment (positiveblue) [inline rpcserver.go:1197]: BTW linter is complaining about not using the pattern

```
if actual.AppMajor != expected.AppMajor {
    return actual.AppMajor > expected.AppMajor
}
```

File: server.go [hunk 4]
```diff
@@ -128,6 +142,14 @@ func (s *Server) Start() error {
 		return nil
 	}
 
+	// Now that we have lnd, lets extract the current version so we can use
+	// this to gate features that we'll use based on the functionality
+	// available.
+	s.lndVersion, err = s.lndServices.Versioner.GetVersion(context.Background())
```
Comment (guggero) [inline server.go:148]: We already have this in `lndServices.Version`, no need to call it again.

File: rpcserver.go (hunk 1 above)
Comment (guggero) [reply rpcserver.go:1192]: It's possible the user has lnd `v0.14.x` but explicitly disables the script enforcement functionality. But then they would just reject any matched orders in the channel acceptor.

We don't need to check build tags though. They are checked on startup already. So we could leave them empty in the `scriptEnforceVersion` variable.

I created a PR for exporting `lndclient.AssertVersionCompatible`, you just need to bump the `lndclient` version to `v0.14.0-7` to get it.

File: rpcserver.go [hunk 5]
```diff
@@ -1185,18 +1186,66 @@ func prepareAndSubmitOrder(ctx context.Context, o order.Order,
...
 
+	// We'll use this channel type selector to pick a channel type based on
+	// the current connected lnd version. This lets us graceful update to
+	// new features as they're available, while still supporting older
+	// versions of lnd.
+	chanTypeSelector := order.WithDefaultChannelType(func() order.ChannelType {
+		// If they didn't specify a value, then we'll select one based
+		// on the version of lnd we detect.
+		if lndVersionIsCompatible(s.server.lndVersion, scriptEnforceVersion) {
+			return order.ChannelTypeScriptEnforced
```
Comment (guggero) [inline rpcserver.go:1236]: Nice use of functional options :100:

**Review by guggero — CHANGES_REQUESTED** [review]
Can be simplified by re-using code from `lndclient`. Other than that LGTM :tada:

File: rpcserver.go (hunk 1 above)
Comment (Roasbeef) [reply rpcserver.go:1192]: Thanks for that PR @guggero! Using that function directly now instead of copying things over.

File: rpcserver.go (hunk 3 above)
Comment (Roasbeef) [reply rpcserver.go:1197]: Should be gone now that the newly exported `lndclient` method is used instead.

File: server.go (hunk 4 above)
Comment (Roasbeef) [reply server.go:148]: Nice, I must've missed that when initially looking over the `godoc`.

**Review by guggero — APPROVED** [review]
LGTM :tada:

**Review by positiveblue — APPROVED** [review]

**Comment by guggero** [comment]
The linter complains about the naming, will fix in separate PR.

**Pull request merged**